nexus/src/app/instance_platform.rs

```diff
@@ -0,0 +1,481 @@
+//! sled to another. To facilitate this, Propolis's migration-request API allows
+//! the caller to pass a list of "replacement" components that the target should
+//! substitute into the specification it receives from the source. Substitutions
+//! are done by component name, so this module must make sure to use a
+//! consistent naming scheme for any components that might be replaced during a
+//! migration. Since (at this writing) the only substitutable components are
+//! backends, this is easily done by using component IDs as backend names, as
+//! described above.
+
+use std::collections::{BTreeMap, HashMap};
+
+use crate::app::instance::InstanceRegisterReason;
+use crate::cidata::InstanceCiData;
+
+use nexus_db_queries::db;
+use nexus_types::identity::Resource;
+use omicron_common::api::external::Error;
+use omicron_common::api::internal::shared::NetworkInterface;
+use sled_agent_client::types::{
+    BlobStorageBackend, Board, BootOrderEntry, BootSettings, Chipset,
+    ComponentV0, CrucibleStorageBackend, I440Fx, InstanceSpecV0, NvmeDisk,
+    PciPath, QemuPvpanic, SerialPort, SerialPortNumber, SpecKey, VirtioDisk,
+    VirtioNetworkBackend, VirtioNic, VmmSpec,
+};
+use uuid::Uuid;
+
+/// Constants and functions used to assign names to assorted VM components.
+mod component_names {
+    pub(super) const COM1: &'static str = "com1";
+    pub(super) const COM2: &'static str = "com2";
+    pub(super) const COM3: &'static str = "com3";
+    pub(super) const COM4: &'static str = "com4";
+    pub(super) const PVPANIC: &'static str = "pvpanic";
+    pub(super) const BOOT_SETTINGS: &'static str = "boot-settings";
+    pub(super) const CLOUD_INIT_DEVICE: &'static str = "cloud-init-dev";
+    pub(super) const CLOUD_INIT_BACKEND: &'static str = "cloud-init-backend";
+
+    /// Given an object ID, derives a name for the "device" half of the
+    /// device/backend component pair that describes that object.
+    pub(super) fn device_name_from_id(id: &uuid::Uuid) -> String {
+        format!("{id}:device")
+    }
+}
+
+enum PciDeviceKind {
+    Disk,
+    Nic,
+    CloudInitDisk,
+}
+
+impl std::fmt::Display for PciDeviceKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Self::Disk => "disk",
+                Self::Nic => "network interface",
+                Self::CloudInitDisk => "cloud-init data disk",
+            }
+        )
+    }
+}
+
+/// Computes the PCI attachment point for a device of the given `kind` which
+/// carries the supplied `logical_slot` in its database record.
+///
+/// WARNING: Guests may write their virtual devices' PCI addresses to persistent
+/// storage and expect that those devices will always appear in the same places
+/// when the system is stopped and restarted. Changing these mappings for
+/// existing instances may break them!
+fn get_pci_bdf(
```

Comment:
pretty nitpicky, feel free to disregard me: i would probably drop the "get" from this, i think of "get" as specifically being an accessor/lookup type thing, and the `zero_padded_nvme_serial_from_str` function in this module doesn't have a `get_` prefix...